uploader/uploader.go

```diff
@@ -1,149 +1,127 @@
+	if bucketName != "" {
+		s3, err := NewS3Backend(m, bucketName, pathPrefix, titusInfo.GetIamProfile(), taskID)
+		if err != nil {
+			return nil, err
+		}
+		return NewUploaderWithBackend(s3), nil
+	}
 
-	e.uploaders = append(e.uploaders, collectCopyUploaders(config)...)
-	e.uploaders = append(e.uploaders, collectS3Uploaders(config, m)...)
-	e.uploaders = append(e.uploaders, collectNoopUploaders(config)...)
+	if len(config.CopyUploaders) > 0 {
+		return NewUploaderWithBackend(NewCopyBackend(config.CopyUploaders[0])), nil
+	}
 
-	return e, nil
+	return NewUploaderWithBackend(NewNoopBackend()), nil
 }
 
-// NewUploadersFromUploaderArray creates a new instance of an Uploaders object from a list of Uploader instances
-func NewUploadersFromUploaderArray(uploaders []Uploader) *Uploaders {
-	e := &Uploaders{
-		uploaders: make([]Uploader, len(uploaders)),
-	}
-	copy(e.uploaders, uploaders)
-	return e
+func NewUploaderWithBackend(backend Backend) *Uploader {
+	return &Uploader{backend}
 }
 
-// Performs a parallel upload of all of files in a directory but
-// not its subdirectories. A slice containing the error results for
-// each upload with an error is returned.
-func uploadDir(ctx context.Context, uploader Uploader, local string, remote string, ctypeFunc ContentTypeInferenceFunction) []error {
-	var errs []error
+// Upload of all of files in a directory but not its subdirectories. Performs the uploads in parallel.
+func uploadDir(ctx context.Context, uploader Backend, local string, remote string, ctypeFunc ContentTypeInferenceFunction) error {
+	var errs *multierror.Error
 
 	fi, err := os.Stat(local)
 	if err != nil {
-		return append(errs, err)
+		errs = multierror.Append(errs, err)
 	}
 
+	var wg sync.WaitGroup
+
 	if fi.IsDir() {
 		finfos, err := ioutil.ReadDir(local)
 		if err != nil {
-			return append(errs, err)
-		}
-		// Iterate over each file and upload it
-		for _, finfo := range finfos {
-			if finfo.IsDir() {
-				continue // don't upload subdirs
-			}
-
-			qlocal := path.Join(local, finfo.Name())
-			qremote := path.Join(remote, finfo.Name())
-
-			log.Printf("Uploading log file %s to %s", qlocal, qremote)
-			if err = uploader.Upload(ctx, qlocal, qremote, ctypeFunc); err != nil {
-				uploadErr := fmt.Errorf("Error uploading to %s: %s", qremote, err)
-				log.Printf("%s", uploadErr)
-				errs = append(errs, uploadErr)
+			errs = multierror.Append(errs, err)
+		} else {
+			// Iterate over each file and upload it
+			for _, finfo := range finfos {
+				if finfo.IsDir() {
+					continue // don't upload subdirs
+				}
+
+				wg.Add(1)
+				go func(fi os.FileInfo) {
+					qlocal := path.Join(local, fi.Name())
+					qremote := path.Join(remote, fi.Name())
+
+					log.WithField("local", qlocal).WithField("remote", qremote).Info("uploading")
+					if err = uploader.Upload(ctx, qlocal, qremote, ctypeFunc); err != nil {
+						log.WithField("local", qlocal).WithField("remote", qremote).Error(err)
+						errs = multierror.Append(errs, err)
```

Comment:
Is this thread-safe?  Can multiple goroutines call this at the same time without losing data?

Reply:
Fixed via 3949578.

Reply:
You might want to try: https://github.com/Netflix/titus-executor/blob/master/vpc/service/errgroupish.go